File: contrib/ifunc/chaser.c
/*
 * Recursive pointer-chasing ifunc 
 * Author: Luis E. P.
 *
 * Copyright (C) 2021-2022 Arm Ltd. All rights reserved.
 *
 */

#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>

#include <ucp/api/ucp.h>

#include "pointer_chase.h"

size_t chaser_payload_bound(void *source_args, size_t source_args_size)
{
	(void)source_args;
	return source_args_size;
}


int chaser_payload_init(void *source_args,
						size_t source_args_size,
						void *payload,
						size_t *payload_size)
{
	(void)source_args_size;
	(void)payload_size;

	pc_src_args_t *src = (pc_src_args_t *)source_args;
	pc_src_args_t *dst = (pc_src_args_t *)payload;

	dst->depth = src->depth;
	dst->addr  = src->addr;
	dst->dest  = src->dest;

	return 0;
}


void chaser_main(void *payload, size_t payload_size, void *target_args)
{
	(void)payload_size;

	pc_src_args_t *src_args = (pc_src_args_t *)payload;
	pc_tgt_args_t *tgt_args = (pc_tgt_args_t *)target_args;
	uint64_t heap_aslr;

	uint64_t depth = src_args->depth;
	uint64_t addr  = src_args->addr;
	uint64_t dest  = src_args->dest;

	uint64_t *data = tgt_args->data;
	int this_shard = tgt_args->shard_id;
	int server_qty = tgt_args->machine_qty - 1;
	int shard_size = tgt_args->shard_size;

	int max_index  = server_qty * shard_size;

	uint64_t next;

	if (VERBOSE) {
		printf("\nI'm going to chase these pointers!\n");
		printf("Depth: %lu\n", depth);
		printf("Addr:  %lu\n", addr);
		printf("Dest:  %lu\n", src_args->dest);
		printf("My shard ID is %d\n", this_shard);
	}

	addr = addr % max_index;	// truncate large numbers
	int which_shard = addr / shard_size;

	if (which_shard == this_shard) {	// data is local
		addr = addr % shard_size;		// get address within a shard

		next = data[addr];
		if (VERBOSE) {
			printf("Result is %lu\n", next);
		}

		// Trim down next to fit
		next = next % max_index;

		depth--;

		if (depth == 0) {

			if (VERBOSE) {
				printf("Reached max depth\n");
			}
			// write to original
			heap_aslr = (uintptr_t)tgt_args->heap + tgt_args->aslr_diff[dest];
			uint64_t *result_ptr;
			ucp_ifunc_msg_t result_msg = tgt_args->result_msg;
			ucp_ifunc_msg_get_payload_ptr(&result_msg, (void **)&result_ptr, NULL);

			*result_ptr = next;

			ucp_ifunc_send_nbix(tgt_args->ep[dest], tgt_args->result_msg,
				                heap_aslr, tgt_args->rmt_rkey[dest]);

			tgt_args->flush(tgt_args->wrkr, tgt_args->ep[dest]);

			return;
		}

		// Update values
		src_args->depth = depth;
		src_args->addr  = next;

		which_shard = next / shard_size;

		if (which_shard == this_shard) {
			chaser_main(payload, payload_size, target_args);
			return;
		}

	}

	// forward to other ep

	int destination = which_shard + 1; // because 0 is the client

	pc_src_args_t *msg_args;	// arguments of the message to be sent
	ucp_ifunc_msg_t i_msg = tgt_args->chaser_msg;

	void *tgt_payload;

    ucp_ifunc_msg_get_payload_ptr(&i_msg, &tgt_payload, NULL);
    
    msg_args = (pc_src_args_t *)tgt_payload;
    msg_args->depth = src_args->depth;
    msg_args->addr  = src_args->addr;
    msg_args->dest  = src_args->dest;

    if (VERBOSE) {
		printf("Remote addr:  %lu\n", msg_args->addr);
		printf("Remote dest:  %lu\n", msg_args->dest);
		printf("Remote depth: %lu\n", msg_args->depth);
	}

	heap_aslr = (uintptr_t)tgt_args->heap + tgt_args->aslr_diff[destination];

	ucp_ifunc_send_nbix(tgt_args->ep[destination], tgt_args->chaser_msg,
		                heap_aslr, tgt_args->rmt_rkey[destination]);
	
	tgt_args->flush(tgt_args->wrkr, tgt_args->ep[destination]);
}

File: contrib/ifunc/julia/chaser.jl
##
# Dynamic adaptive pointer chasing application
#
# Author: Valentin Churavy
# Massachussetts Institute of Technology, 2022
#
# Original implementation:
# Author: Luis E. P.
# Copyright (C) 2021-2022 Arm Ltd. All rights reserved.
#
##
using UCX

include("UCXAsap.jl")
import .UCXAsap

include("IFunc.jl")
import .IFunc 

include("unsafe.jl")
include("common.jl")

function bound(source_args::Ptr{Cvoid}, source_args_size::Csize_t)::Csize_t
	return source_args_size
end

function init(source_args::Ptr{Cvoid}, source_args_size::Csize_t, payload::Ptr{Cvoid}, payload_size::Csize_t)::Cint
    src = Base.unsafe_convert(Ptr{SrcArgs}, source_args)
    dst = Base.unsafe_convert(Ptr{SrcArgs}, payload)

    unsafe_setfield!(dst, :depth, unsafe_getfield(src, :depth))
    unsafe_setfield!(dst, :addr, unsafe_getfield(src, :addr))
    unsafe_setfield!(dst, :dest, unsafe_getfield(src, :dest))

    return Cint(0)
end

function main(payload::Ptr{Cvoid}, payload_size::Csize_t, target_args::Ptr{Cvoid})::Cvoid
    src_args = Base.unsafe_convert(Ptr{SrcArgs}, payload)
    tgt_args = Base.unsafe_convert(Ptr{TgtArgs}, target_args)

    depth = unsafe_getfield(src_args, :depth)
    addr  = unsafe_getfield(src_args, :addr)
    dest  = unsafe_getfield(src_args, :dest)%Int64

    data = unsafe_getfield(tgt_args, :data)
    this_shard = unsafe_getfield(tgt_args, :shard_id)
    server_qty = unsafe_getfield(tgt_args, :machine_qty) - 1
    shard_size = unsafe_getfield(tgt_args, :shard_size)
    worker = unsafe_getfield(tgt_args, :wrkr)

    max_index = server_qty * shard_size

    # addr = mod(addr, max_index)
    # @assert addr in 0:(max_index-1)
    which_shard = unsafe_div(addr, shard_size)

    if which_shard == this_shard # data is local
		addr = unsafe_rem(addr, shard_size) # get address within a shard

        next = Base.unsafe_load(data, (addr + 1)%Int64) # Julia is 1-based
        # @static if VERBOSE
        #     @printf("Result is %lu\n", next)
        # end

        # Trim down next to fit
        # next = unsafe_rem(next, max_index)
        # @assert next in 0:(max_index-1)

        depth -= 1

        if depth == 0
            dest += 1 # `dest` is a offset, turn to 1-based
            # write to original
            heap_u64 = reinterpret(UInt64, unsafe_getfield(tgt_args, :heap))
            heap_aslr = UInt64(heap_u64 + Base.unsafe_load(unsafe_getfield(tgt_args, :aslr_diff), dest))

            result_msg = unsafe_getfield(tgt_args, :result_msg)
            result_payload, _ = IFunc.ucp_ifunc_msg_get_payload_ptr(result_msg)

            Base.unsafe_store!(Base.unsafe_convert(Ptr{UInt64}, result_payload), next)

            ep = Base.unsafe_load(unsafe_getfield(tgt_args, :ep), dest)
            rmt_rkey = Base.unsafe_load(unsafe_getfield(tgt_args, :rmt_rkey), dest)

            s = IFunc.ucp_ifunc_send_nbix(ep, result_msg, heap_aslr, rmt_rkey)
            
            if s != UCX.API.UCS_OK
                flush = unsafe_getfield(tgt_args, :flush)
                assume(flush != C_NULL)
                ccall(flush, UCX.API.ucs_status_t, (UCX.API.ucp_worker_h, UCX.API.ucp_ep_h), worker, ep)
            end
            return
        end

        # update values
        unsafe_setfield!(src_args, :depth, depth)
        unsafe_setfield!(src_args, :addr, next)

        which_shard = unsafe_div(next, shard_size)
        if which_shard == this_shard 
            # shortcircuit
            main(payload, payload_size, target_args)
            return
        end

    end

    # forward to other ep
    destination = (which_shard%Int64 + 1) + 1 # client is 1, first server is 2
    chaser_msg  = unsafe_getfield(tgt_args, :chaser_msg)
    tgt_payload, _ = IFunc.ucp_ifunc_msg_get_payload_ptr(chaser_msg)

    msg_args = Base.unsafe_convert(Ptr{SrcArgs}, tgt_payload)
    unsafe_setfield!(msg_args, :depth, depth)
    unsafe_setfield!(msg_args, :addr, addr)
    unsafe_setfield!(msg_args, :dest, dest % UInt64)

    # @static if VERBOSE
    #     @printf("Remote addr:  %lu\n", msg_args.addr)
	# 	@printf("Remote dest:  %lu\n", msg_args.dest)
	# 	@printf("Remote depth: %lu\n", msg_args.depth)
    # end

    heap_u64 = reinterpret(UInt64, unsafe_getfield(tgt_args, :heap))
    heap_aslr = UInt64(heap_u64 + Base.unsafe_load(unsafe_getfield(tgt_args, :aslr_diff), destination))

    ep = Base.unsafe_load(unsafe_getfield(tgt_args, :ep), destination)
    rmt_rkey = Base.unsafe_load(unsafe_getfield(tgt_args, :rmt_rkey), destination)

    s = IFunc.ucp_ifunc_send_nbix(ep, chaser_msg, heap_aslr, rmt_rkey)

    if s != UCX.API.UCS_OK
        flush = unsafe_getfield(tgt_args, :flush)
        assume(flush != C_NULL)
        ccall(flush, UCX.API.ucs_status_t, (UCX.API.ucp_worker_h, UCX.API.ucp_ep_h), worker, ep)
    end

    return nothing
end

const fns = (
    (bound, Tuple{Ptr{Cvoid}, Csize_t}),
    (init, Tuple{Ptr{Cvoid}, Csize_t, Ptr{Cvoid}, Csize_t}),
    (main, Tuple{Ptr{Cvoid}, Csize_t, Ptr{Cvoid}})
)

IFunc.Compiler.emit(fns, "jchaser", "$(Sys.ARCH)-unknown-linux-gnu")
